File: backend/routes/quality/accreditation.py
from flask import Blueprint, jsonify, request
from functools import wraps
from supabase_client import get_supabase

# Create blueprint
quality_accreditation_bp = Blueprint('quality_accreditation', __name__)

@quality_accreditation_bp.route('/accreditation/readiness', methods=['GET', 'OPTIONS'])
def get_readiness_score():
    """Get accreditation readiness score from quality_accreditation table"""
    try:
        supabase = get_supabase()

        # Get the latest report
        result = supabase.table('quality_accreditation').select('*').order('report_date', desc=True).limit(1).execute()

        if not result.data:
            return jsonify({
                'success': True,
                'data': {
                    'overall_score': 0,
                    'readiness_level': 'poor',
                    'criteria_scores': {},
                    'department_scores': {}
                }
            })

        report = result.data[0]
        score = float(report.get('score', 0))
        
        # Determine readiness level based on score
        if score >= 90:
            level = 'excellent'
        elif score >= 80:
            level = 'good'
        elif score >= 70:
            level = 'average'
        else:
            level = 'poor'

        # Mock criteria scores for the radar chart
        criteria_scores = {
            'Teaching & Learning': score * 0.9,
            'Research & Innovation': score * 0.8,
            'Infrastructure': score * 0.95,
            'Governance': score * 0.85,
            'Student Support': score * 0.9,
            'Extension Activities': score * 0.75
        }

        # Mock department scores for the bar chart
        department_scores = {
            'Computer Science': score * 0.95,
            'Mechanical Engineering': score * 0.9,
            'Electrical Engineering': score * 0.85,
            'Civil Engineering': score * 0.8,
            'Business Administration': score * 0.88
        }

        readiness_score = {
            'overall_score': score,
            'readiness_level': level,
            'criteria_scores': criteria_scores,
            'department_scores': department_scores
        }

        return jsonify({
            'success': True,
            'data': readiness_score
        })
    except Exception as e:
        return jsonify({
            'success': False,
            'error': str(e)
        }), 500

@quality_accreditation_bp.route('/accreditation/analytics', methods=['GET', 'OPTIONS'])
def get_accreditation_analytics():
    """Get accreditation analytics"""
    try:
        supabase = get_supabase()
        result = supabase.table('quality_accreditation').select('*').order('report_date', desc=True).execute()
        data = result.data or []
        
        score_trends = []
        for r in reversed(data[:6]): # Last 6 reports
            score_trends.append({
                'date': r.get('report_date', ''),
                'score': float(r.get('score', 0))
            })
            
        dept_readiness = []
        for r in data[:5]: # Last 5 unique departments maybe
            dept_readiness.append({
                'department': r.get('department', 'General'),
                'readiness_score': float(r.get('score', 0))
            })
        
        level_counts = {'excellent': 0, 'good': 0, 'average': 0, 'poor': 0}
        for r in data:
            score = float(r.get('score', 0))
            if score >= 90:
                level = 'excellent'
            elif score >= 80:
                level = 'good'
            elif score >= 70:
                level = 'average'
            else:
                level = 'poor'
            level_counts[level] += 1
                
        readiness_distribution = [{'level': k, 'count': v} for k, v in level_counts.items()]

        analytics = {
            'score_trends': score_trends,
            'department_readiness': dept_readiness,
            'readiness_distribution': readiness_distribution
        }
        
        return jsonify({
            'success': True,
            'data': analytics
        })
    except Exception as e:
        return jsonify({
            'success': False,
            'error': str(e)
        }), 500

@quality_accreditation_bp.route('/accreditation', methods=['GET', 'OPTIONS'])
def get_accreditation():
    """Get accreditation information from Supabase"""
    try:
        supabase = get_supabase()

        # Get pagination parameters
        page = int(request.args.get('page', 1))
        limit = int(request.args.get('limit', 10))
        offset = (page - 1) * limit

        # Get filter parameters
        search = request.args.get('search', '')

        # Build query
        query = supabase.table('quality_accreditation').select('*', count='exact')

        # Apply filters
        if search:
            query = query.or_(f"report_type.ilike.%{search}%,department.ilike.%{search}%")

        # Execute query with pagination
        result = query.range(offset, offset + limit - 1).execute()
        total_count = result.count if hasattr(result, 'count') else 0

        # Map database fields to frontend expected fields
        accreditation = []
        for record in (result.data or []):
            score = float(record.get('score', 0))
            mapped_record = {
                'id': record.get('report_id'),
                'body': record.get('report_type', 'NAAC'),
                'program': record.get('department', 'Institutional'),
                'status': 'completed',
                'valid_until': '', 
                'grade': 'A+' if score >= 90 else
                        'A' if score >= 80 else
                        'B+' if score >= 70 else 'B'
            }
            accreditation.append(mapped_record)

        total_pages = (total_count + limit - 1) // limit if total_count > 0 else 1

        return jsonify({
            'success': True,
            'data': accreditation,
            'pagination': {
                'currentPage': page,
                'totalPages': total_pages,
                'totalItems': total_count,
                'limit': limit
            }
        })

    except Exception as e:
        return jsonify({
            'success': False,
            'error': str(e)
        }), 500

@quality_accreditation_bp.route('/accreditation/reports', methods=['GET', 'OPTIONS'])
def get_accreditation_reports():
    """Get accreditation reports with pagination from quality_accreditation table"""
    try:
        supabase = get_supabase()
        page = int(request.args.get('page', 1))
        limit = int(request.args.get('limit', 10))
        offset = (page - 1) * limit

        result = supabase.table('quality_accreditation').select('*', count='exact').range(offset, offset + limit - 1).execute()
        total_count = result.count if hasattr(result, 'count') else 0

        reports = []
        for record in (result.data or []):
            score = float(record.get('score', 0))
            if score >= 90:
                level = 'excellent'
            elif score >= 80:
                level = 'good'
            elif score >= 70:
                level = 'average'
            else:
                level = 'poor'
                
            reports.append({
                'id': record['report_id'],
                'accreditation_body': record['report_type'],
                'academic_year': 'Current',
                'overall_score': score,
                'readiness_level': level,
                'generated_date': record['report_date'],
                'status': 'completed',
                'recommendations': record.get('recommendations', [])
            })

        total_pages = (total_count + limit - 1) // limit if total_count > 0 else 1

        return jsonify({
            'success': True,
            'data': reports,
            'pagination': {
                'page': page,
                'limit': limit,
                'total': total_count,
                'totalPages': total_pages
            }
        })
    except Exception as e:
        return jsonify({
            'success': False,
            'error': str(e)
        }), 500

@quality_accreditation_bp.route('/accreditation/reports', methods=['POST', 'OPTIONS'])
def generate_accreditation_report():
    """Generate accreditation report"""
    try:
        data = request.get_json()
        
        # Mock report generation
        new_report = {
            'id': 999,
            'accreditation_body': data.get('accreditation_body', 'NAAC'),
            'academic_year': data.get('academic_year', '2024'),
            'overall_score': 85,
            'readiness_level': 'good',
            'generated_date': '2024-01-07',
            'status': 'draft'
        }
        
        return jsonify({
            'success': True,
            'data': new_report,
            'message': 'Accreditation report generated successfully'
        })
    except Exception as e:
        return jsonify({
            'success': False,
            'error': str(e)
        }), 500

File: Updated_CMS-main/backend/controllers/quality/accreditation.js
const { supabase } = require('../../supabase_client');

// Helper function to calculate readiness score
async function calculateReadinessScore() {
  try {
    // Get faculty performance data
    const { data: faculty } = await supabase
      .from('quality_facultyperformance')
      .select('performance_rating, department, research_papers');

    // Get audit compliance data
    const { data: audits } = await supabase
      .from('quality_audits')
      .select('compliance_score, department');

    // Get policy compliance data
    const { data: policies } = await supabase
      .from('quality_policy')
      .select('compliance_status');

    // Calculate criteria scores
    const criteriaScores = {
      'Curriculum': calculateFacultyScore(faculty),
      'Teaching-Learning': calculateFacultyScore(faculty) * 0.9,
      'Research': calculateResearchScore(faculty),
      'Infrastructure': 85, // Mock score
      'Student Support': 80, // Mock score (grievances table missing)
      'Governance': calculatePolicyScore(policies),
      'Innovative Practices': 80, // Mock score
    };

    // Calculate department scores
    const departmentScores = {};
    const departments = [...new Set(faculty?.map(f => f.department) || [])];
    
    departments.forEach(dept => {
      const deptFaculty = faculty?.filter(f => f.department === dept) || [];
      const deptAudits = audits?.filter(a => a.department === dept) || [];
      
      departmentScores[dept] = Math.round(
        (calculateFacultyScore(deptFaculty) + 
         calculateAuditScore(deptAudits)) / 2
      );
    });

    // Calculate overall score
    const overallScore = Math.round(
      Object.values(criteriaScores).reduce((sum, score) => sum + score, 0) / 
      Object.keys(criteriaScores).length
    );

    return {
      overall_score: overallScore,
      criteria_scores: criteriaScores,
      department_scores: departmentScores
    };
  } catch (error) {
    console.error('Error calculating readiness score:', error);
    return {
      overall_score: 0,
      criteria_scores: {},
      department_scores: {}
    };
  }
}

// Helper functions for score calculations
function calculateFacultyScore(faculty) {
  if (!faculty || faculty.length === 0) return 0;
  const total = faculty.reduce((sum, f) => sum + (f.performance_rating || 0), 0);
  return Math.round(total / faculty.length);
}

function calculateResearchScore(faculty) {
  if (!faculty || faculty.length === 0) return 0;
  const total = faculty.reduce((sum, f) => sum + (f.research_papers || 0), 0);
  return Math.min(100, Math.round(total / faculty.length * 10));
}

function calculatePolicyScore(policies) {
  if (!policies || policies.length === 0) return 0;
  const compliant = policies.filter(p => p.compliance_status === 'Compliant').length;
  return Math.round((compliant / policies.length) * 100);
}

function calculateAuditScore(audits) {
  if (!audits || audits.length === 0) return 0;
  const total = audits.reduce((sum, a) => sum + (a.compliance_score || 0), 0);
  return Math.round(total / audits.length);
}

function getReadinessLevel(score) {
  if (score >= 90) return 'excellent';
  if (score >= 75) return 'good';
  if (score >= 60) return 'average';
  return 'poor';
}

function generateRecommendations(readinessScore) {
  const recommendations = [];
  
  Object.entries(readinessScore.criteria_scores).forEach(([criteria, score]) => {
    if (score < 75) {
      recommendations.push(`Improve ${criteria} performance (Current: ${score}%)`);
    }
  });
  
  if (readinessScore.overall_score < 80) {
    recommendations.push('Focus on overall institutional development');
  }
  
  return recommendations;
}

function identifyStrengths(readinessScore) {
  const strengths = [];
  
  Object.entries(readinessScore.criteria_scores).forEach(([criteria, score]) => {
    if (score >= 85) {
      strengths.push(`${criteria} (${score}%)`);
    }
  });
  
  return strengths;
}

function identifyWeaknesses(readinessScore) {
  const weaknesses = [];
  
  Object.entries(readinessScore.criteria_scores).forEach(([criteria, score]) => {
    if (score < 70) {
      weaknesses.push(`${criteria} (${score}%)`);
    }
  });
  
  return weaknesses;
}

// SUB-MODULE 6: ACCREDITATION READINESS REPORTS
exports.getReadinessScore = async (req, res) => {
  try {
    const readinessScore = await calculateReadinessScore();
    
    res.json({
      success: true,
      data: readinessScore
    });
  } catch (error) {
    console.error('Error calculating readiness score:', error);
    res.status(500).json({
      success: false,
      error: 'Failed to calculate readiness score'
    });
  }
};

exports.getAllReports = async (req, res) => {
  try {
    const { page = 1, limit = 10, accreditation_body, status } = req.query;
    
    let query = supabase
      .from('quality_accreditation')
      .select('*', { count: 'exact' })
      .order('report_date', { ascending: false });

    if (accreditation_body) query = query.eq('report_type', accreditation_body);

    const from = (page - 1) * limit;
    const to = from + limit - 1;

    const { data, error, count } = await query.range(from, to);

    if (error) throw error;

    const mappedData = data ? data.map(report => ({
      id: report.report_id,
      accreditation_body: report.report_type,
      generated_date: report.report_date,
      overall_score: report.score,
      status: 'completed',
      recommendations: [report.recommendations],
      department: report.department
    })) : [];

    res.json({
      success: true,
      data: mappedData,
      pagination: {
        page: parseInt(page),
        limit: parseInt(limit),
        total: count || 0,
        totalPages: Math.ceil((count || 0) / limit)
      }
    });
  } catch (error) {
    console.error('Error fetching accreditation reports:', error);
    res.status(500).json({
      success: false,
      error: 'Failed to fetch accreditation reports'
    });
  }
};

exports.generateReport = async (req, res) => {
  try {
    const { accreditation_body, academic_year, department } = req.body;
    
    // Calculate readiness score and generate report
    const readinessScore = await calculateReadinessScore();
    
    const reportData = {
      report_type: accreditation_body,
      department: department || 'Institutional',
      report_date: new Date().toISOString().split('T')[0],
      score: readinessScore.overall_score,
      recommendations: generateRecommendations(readinessScore).join(', ')
    };

    const { data, error } = await supabase
      .from('quality_accreditation')
      .insert([reportData])
      .select();

    if (error) throw error;

    res.status(201).json({
      success: true,
      data: {
        id: data[0].report_id,
        accreditation_body: data[0].report_type,
        generated_date: data[0].report_date,
        overall_score: data[0].score,
        recommendations: [data[0].recommendations],
        department: data[0].department
      },
      message: 'Accreditation report generated successfully'
    });
  } catch (error) {
    console.error('Error generating accreditation report:', error);
    res.status(500).json({
      success: false,
      error: 'Failed to generate accreditation report'
    });
  }
};

exports.getAccreditationAnalytics = async (req, res) => {
  try {
    // Get historical reports for trend analysis
    const { data: reports } = await supabase
      .from('quality_accreditation')
      .select('*')
      .order('report_date', { ascending: true });

    // Score trends
    const scoreTrends = reports ? 
      reports.map(report => ({
        date: report.report_date,
        score: report.score,
        body: report.report_type
      })) : [];

    // Department-wise readiness
    const { data: faculty } = await supabase
      .from('quality_facultyperformance')
      .select('department, performance_rating');

    const departmentReadiness = faculty ? 
      faculty.reduce((acc, f) => {
        const dept = acc.find(item => item.department === f.department);
        if (dept) {
          dept.total += f.performance_rating;
          dept.count += 1;
        } else {
          acc.push({
            department: f.department,
            total: f.performance_rating,
            count: 1
          });
        }
        return acc;
      }, []).map(item => ({
        department: item.department,
        readiness_score: Math.round(item.total / item.count)
      })) : [];

    // Readiness distribution
    const readinessDistribution = [
      { level: 'Excellent', count: 0, range: '90-100' },
      { level: 'Good', count: 0, range: '75-89' },
      { level: 'Average', count: 0, range: '60-74' },
      { level: 'Poor', count: 0, range: '0-59' }
    ];

    if (reports) {
      reports.forEach(report => {
        if (report.score >= 90) readinessDistribution[0].count++;
        else if (report.score >= 75) readinessDistribution[1].count++;
        else if (report.score >= 60) readinessDistribution[2].count++;
        else readinessDistribution[3].count++;
      });
    }

    const analytics = {
      score_trends: scoreTrends,
      department_readiness: departmentReadiness,
      readiness_distribution: readinessDistribution
    };

    res.json({
      success: true,
      data: analytics
    });
  } catch (error) {
    console.error('Error fetching accreditation analytics:', error);
    res.status(500).json({
      success: false,
      error: 'Failed to fetch accreditation analytics'
    });
  }
};
